### ISSUE TITLE
Metal OOM crash loop after GPU Hang — driver memory leak not recoverable without reboot

### ISSUE THREAD

**Issue by Neo2025new** [jundot/omlx#15]
## Environment

- **Hardware**: Mac Studio M3 Ultra (512GB Unified Memory)
- **macOS**: Darwin 25.3.0 (Sequoia)
- **oMLX**: v0.1.2
- **Model**: MiniMax-M2.5-MLX-8bit (estimated 237.61GB)
- **Config**: `--max-model-memory 450GB --paged-ssd-cache-dir ~/.omlx/cache --paged-ssd-cache-max-size 100GB`

## Summary

After a Metal GPU Hang during inference, oMLX enters an unrecoverable crash loop. The Metal driver leaks GPU memory across crash-restart cycles, eventually causing `kIOGPUCommandBufferCallbackErrorOutOfMemory` on every model load attempt. **Only a full system reboot clears the leaked memory.**

## Crash Timeline (2026-02-15)

```
21:01  Load MiniMax-M2.5-MLX-8bit → success (237.61GB)
21:02  GPU Hang during inference
       [METAL] Command buffer execution failed: Caused GPU Hang Error
         (kIOGPUCommandBufferCallbackErrorHang)
       [METAL] Command buffer execution failed: Discarded (victim of GPU error/recovery)
         (kIOGPUCommandBufferCallbackErrorInnocentVictim)
       → oMLX crashes, LaunchAgent restarts

21:13  Restart → model loads OK → runs for ~33 minutes
21:46  Inference deadlocks (process alive, CPU 0.1%, 5 ESTABLISHED connections, no response)
       → forced restart via launchctl kickstart -k

21:47  Load model → success → IMMEDIATE Metal OOM
       [METAL] Command buffer execution failed: Insufficient Memory
         (kIOGPUCommandBufferCallbackErrorOutOfMemory)

21:49  Auto-restart → load model → success → IMMEDIATE Metal OOM again
21:50  Auto-restart → server running, model not loaded (will OOM on next request)
```

**6 Metal errors total in one session.** Three distinct error types.

## Root Cause Analysis

After the crash loop, system memory shows:

| Metric | Value | Note |
|--------|-------|------|
| Wired (non-evictable) | **249 GB** | 🚨 Abnormal |
| Free | 225 GB | |
| All user processes RSS combined | **~4 GB** | oMLX itself is only 125MB |

**~245GB of wired memory is unaccounted for** — not attributable to any user process. This is leaked Metal/IOKit driver memory from the GPU Hang recovery + crash cycles. The Metal driver does not fully release `IOGPUResource` references when oMLX terminates after a GPU error.

With 249GB wired, loading the 237GB model leaves only ~26GB for Metal compute buffers (attention, KV cache blocks, etc.), triggering OOM immediately after model load.

## Key Observations

1. **The 4-bit model (125GB) also fails** once enough driver memory has leaked — confirming it's not a model size issue
2. **The deadlock at 21:46** (process alive but not responding, CPU 0.1%) may be a related but distinct issue — possibly a hung Metal command buffer that never completes or times out
3. **Paged SSD Cache was active** (`initial_blocks=256, max_blocks=100000, block_size=256 tokens`) — the SSD paging worked correctly in earlier sessions; this is specifically triggered by GPU Hang recovery

## Suggested Improvements

1. **Graceful Metal resource cleanup on crash**: Trap SIGTERM/SIGABRT and explicitly release Metal device, command queues, and buffers before exit. This won't fix the driver leak but may reduce it.

2. **Metal health check before model load**: After restart, probe Metal with a small allocation to verify GPU memory is actually available before committing to a 237GB model load.

3. **Inference timeout / deadlock detection**: The 21:46 deadlock (33 minutes of running → silent hang) could benefit from a watchdog that detects when inference requests stop completing and triggers a graceful restart.

4. **Consider `--max-model-memory` headroom warning**: When model estimate + a safety margin (e.g., 50GB for Metal compute) exceeds system memory, warn the user at startup.

## Reproduction

1. Load a large model (>200GB) on 512GB Apple Silicon
2. Run sustained inference until a GPU Hang occurs (may take hours)
3. After the first crash-restart, attempt inference again
4. Observe Metal OOM on subsequent load attempts
5. `vm_stat` will show wired pages consuming far more than any process accounts for

## Logs

Full error log segment:
```
2026-02-15 21:50:02,581 - omlx.engine_pool - INFO - Loaded model: MiniMax-M2.5-MLX-8bit (estimated: 237.61GB, total: 237.61GB)
libc++abi: terminating due to uncaught exception of type std::runtime_error: [METAL] Command buffer execution failed: Insufficient Memory (00000008:kIOGPUCommandBufferCallbackErrorOutOfMemory)
```

**Comment by 0xNeoTian** [comment]
## Update: Inference deadlock pattern (distinct from Metal OOM)

After rebooting and clearing the Metal driver leak, I observed a **separate recurring issue**: inference gradually becomes unresponsive after running for ~15-30 minutes, without any Metal errors in the log.

### Symptoms
- `/v1/models` responds normally (HTTP 200)
- `/v1/chat/completions` either:
  - Times out completely (10s+)
  - Responds extremely slowly (25s for 3 tokens, normally 2s)
- Process alive, RSS normal (226GB), CPU 0.1-55% (variable)
- No Metal errors, no OOM, no crash in stderr
- ESTABLISHED connections accumulate from clients
- `ps -M` shows all threads in Sleep state

### Timeline (post-reboot, same session)
```
22:05:25 - Chat completion: 10 tokens in 2.03s (4.9 tok/s)  ← normal
22:05:37 - Chat completion: 10 tokens in 2.04s (4.9 tok/s)  ← normal
22:17:xx - Inference timeout / 25s for 3 tokens              ← degraded
```

~12 minutes between last successful fast inference and degradation.

### Workaround
`launchctl kickstart -k` restarts oMLX, model reloads in ~35s, inference returns to normal (2.9s, 11.6 tok/s).

### Hypothesis
Possibly related to:
1. **Paged SSD Cache interactions**: 317 cached blocks (19.2GB), block scan + prefix matching may degrade over time
2. **Continuous batching scheduler**: With persistent connections from OpenClaw, the scheduler may enter a state where new requests are queued but never scheduled
3. **MLX Metal compute thread starvation**: All 4 Metal threads show ~8s STIME (system time) — possibly blocked on GPU synchronization

This is a **different issue** from the Metal OOM crash loop (which was driver memory leak). This one leaves no trace in logs — the process just silently stops processing inference.

**Comment by 0xNeoTian** [comment]
## Update: Inference deadlock root cause identified — Paged SSD Cache

### Finding

After extensive debugging, I've identified that the **Paged SSD Cache** is the root cause of the inference deadlock described in my previous comment. Disabling it completely resolved the issue.

### Evidence

**With SSD cache enabled** (`--paged-ssd-cache-dir ~/.omlx/cache`, 24GB / 402 files):
- Inference deadlocks every 2-15 minutes
- Process alive (CPU 45-59%), accepts connections, but never completes inference
- No errors in logs — completely silent hang
- Speed degrades progressively: 11.6 → 2.0 tok/s before full deadlock
- Happened consistently across multiple restarts and both 4-bit and 8-bit models

**With SSD cache disabled** (`--no-cache` + `settings.json cache.enabled: false`):
- Stable inference over extended period, no deadlocks
- Consistent 1.8-4.7s response times, 11+ tok/s
- Same process (PID) stays alive without crashes

### Suspected mechanism

The SSD cache I/O operations (reading/writing 402 block files, 24GB total) appear to contend with Metal GPU compute, causing the scheduler to block indefinitely waiting for cache block operations that never complete.

### Reproduction

1. Start oMLX with `--paged-ssd-cache-dir ~/.omlx/cache` on Apple Silicon with a large model (237GB 8-bit)
2. Send several inference requests over 5-15 minutes
3. Observe: response times degrade, then requests hang indefinitely
4. Compare: restart with `--no-cache` — stable indefinitely

### Notes on `--no-cache` flag

The `--no-cache` CLI flag correctly sets `paged_ssd_cache_dir = None` in `cli.py`, but `settings.json` `cache.enabled: true` may still cause cache initialization through the settings pipeline. I had to set **both** `--no-cache` in CLI args AND `cache.enabled: false` in `settings.json` to fully disable SSD cache. This might be a config priority bug worth investigating.

### Environment
- Mac Studio M3 Ultra, 512GB RAM
- MiniMax-M2.5-MLX-8bit (237GB)
- macOS 15.4
- oMLX latest (editable install from source)

**Comment by jundot** [comment]
I've also tested tens of millions of tokens with Minimax 2.1 through Claude Code, and this is an issue I'm seeing for the first time! Unfortunately, I don't have much free time on weekdays, so it might take a bit longer. If possible, it would be helpful if you could let me know the name of the coding agent you're using.

**Comment by 0xNeoTian** [comment]
Hi @jundot, thanks for looking into this!

The coding agent is **Claude Code** (Anthropic's official CLI). It connects to oMLX via the native Anthropic API endpoint. However, this issue is **not agent-specific** — I reproduced the deadlock consistently with simple `curl` requests as well.

**Good news: I've identified the root cause and implemented a fix.** Submitting a PR now.

### Root Cause

`save_block()` and `load_block()` in `paged_ssd_cache.py` perform **synchronous disk I/O** (`mx.save_safetensors()` / `mx.load()`) while holding `self._lock` (RLock) on the inference thread. After the SSD cache accumulates enough blocks (typically 2-15 minutes of inference), these I/O operations block `scheduler.step()` long enough to cause a deadlock — inference hangs indefinitely with no error in logs.

### Fix: Async I/O with Metal-safe writes

1. **Write path**: Main thread materializes arrays via `mx.eval()` + writes temp file via `mx.save_safetensors()` → background daemon thread does atomic `os.rename()` → `save_block()` returns immediately
2. **Read path**: Check `_pending_writes` in-memory buffer first (zero I/O) → fallback to `ThreadPoolExecutor` + `future.result(timeout=5s)` → timeout = cache miss (graceful degradation)
3. **Key constraint**: All `mx.*` / Metal operations stay on the main thread — background thread only does `os.rename()` (Metal command buffers are not thread-safe; calling `mx.save_safetensors()` from a background thread causes `commit command buffer with uncommitted encoder` assertion crash)

### Test Results
- 54 unit tests passing (including 9 new tests for async I/O)
- **100/100 live inference requests over 28+ minutes, zero deadlocks**
- MiniMax-M2.5-MLX-8bit (237GB), Mac Studio M3 Ultra 512GB

PR incoming.

---

**Note on the Metal GPU Hang / OOM crash loop** (the original issue in this thread): That is a separate Apple Silicon driver-level issue — after a GPU Hang, the Metal driver leaks wired memory across process restarts, only recoverable by rebooting. The SSD cache deadlock fix does not address the Metal driver leak, but it eliminates the most common trigger for the initial GPU Hang (the I/O-induced stall that could destabilize Metal scheduling).

**Comment by jundot** [comment]
The SSD cache deadlock fix in #16 addresses the primary trigger for this crash loop. The repeated inference hangs caused forced restarts, which accumulated Metal driver memory leaks leading to OOM.

The initial GPU hang at 21:02 (one minute after model load) may be a separate concern, since the SSD cache deadlock typically manifests after 2-15 minutes of operation. I'll revisit this if similar Metal errors recur after the fix is merged.
